File: 200-number-of-islands/number-of-islands.java
class Solution {
    boolean[][] visited;
    int m;
    int n;

    public int numIslands(char[][] grid) {
        int ans = 0;
        // dfs to traverse all the islands
        this.m = grid.length;
        this.n = grid[0].length;
        this.visited = new boolean[m][n];

        for (int i = 0; i < this.m; i++) {
            for (int j = 0; j < this.n; j++) {
                if (!visited[i][j] && grid[i][j] == '1') {
                    dfs(i, j, grid);
                    ans++;
                }
            }
        }
        return ans;
    }

    public void dfs(int i, int j, char[][] grid) {
        // base cases
        if (i < 0 || i >= this.m || j < 0 || j >= this.n || grid[i][j] == '0' || visited[i][j]) {
            return;
        }

        visited[i][j] = true;

        // 4 directions
        dfs(i + 1, j, grid);
        dfs(i - 1, j, grid);
        dfs(i, j + 1, grid);
        dfs(i, j - 1, grid);
    }
}

File: 200-number-of-islands/number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let ans = 0;
    let m = grid.length;
    let n = grid[0].length;
    // graph 
    for(let r=0; r < m; r++){
        for(let c=0; c<n; c++){
            if(grid[r][c] === '1') {
                dfs(r, c);
                ans++;
            }
        }
    }
    // dfs
    function dfs(r, c){
        if(r < 0 || c < 0 || r>= m || c >=n || grid[r][c] === '0'){
            return;
        }
        grid[r][c] = '0';
        dfs(r + 1, c);
        dfs(r - 1, c);
        dfs(r, c - 1);
        dfs(r, c + 1);
    }

    // 
    return ans;
};